### PR TITLE
update mdMessages directive

### PR DESCRIPTION
**Pull request by vincentpalita**
- renamed mdMessages to camelcase.
- receives a reference to formControlName to match new form controls.

Before:

```
[md-messages]="control"
md-message="required"
```

After: 

```
mdMessages="control"
mdMessage="required"
```
- added unit test for md-invalid/valid class on mdMessages container.
- added example for required field in input demo on postalCode field.

### CODE REVIEW

**Comment by vincentpalita** [comment]
I removed my previous pull request, because I accidentally included commits from master branch.
Now it should be more straightforward to review it.
Thanks.

File: modules/site/src/app/examples/input/input-basic-usage.component.ts [hunk 1]
```diff
@@ -22,15 +22,26 @@ export class InputBasicUsageComponent {
...
+  form:FormGroup;
+  postalCode:FormControl;
+
   states = [
     'AL', 'AK', 'AZ', 'AR', 'CA', 'CO', 'CT', 'DE', 'FL', 'GA', 'HI', 'ID', 'IL',
     'IN', 'IA', 'KS', 'KY', 'LA', 'ME', 'MD', 'MA', 'MI', 'MN', 'MS', 'MO', 'MT',
     'NE', 'NV', 'NH', 'NJ', 'NM', 'NY', 'NC', 'ND', 'OH', 'OK', 'OR', 'PA', 'RI',
     'SC', 'SD', 'TN', 'TX', 'UT', 'VT', 'VA', 'WA', 'WV', 'WI', 'WY'
-  ].map(function(state) { return {abbrev: state}; })
+  ].map(function (state) {
```
Comment (Gregcop1) [inline modules/site/src/app/examples/input/input-basic-usage.component.ts]:
```
].map((state) => {abbrev: state});
```

and even shorter:
```
].map((abbrev) => {abbrev});
```

File: modules/site/src/main.ts [hunk 2]
```diff
@@ -13,6 +13,7 @@ if (environment.production) {
 bootstrap(SiteAppComponent, [
   ...AppRouterProviders, ...DEMO_PROVIDERS, ...HTTP_PROVIDERS, ...MATERIAL_BROWSER_PROVIDERS,
   {provide: LocationStrategy, useClass: HashLocationStrategy},
+  {provide: QueryList, useClass: QueryList},
```
Comment (Gregcop1) [inline modules/site/src/main.ts]: why setting it globally instead of locally ?

File: src/components/form/messages.ts [hunk 3]
```diff
@@ -33,92 +33,59 @@ export class MdMessage {
...
-      let prop = <string>this.property;
-      let group = this.form.control;
-      let ctrl: AbstractControl = group.controls[prop];
-      errors = ctrl.errors;
-    }
+    let errors:{[errorKey:string]:string} = null;
+    errors = this.control.errors;
     if (errors) {
-      this.messages.toArray().forEach((m: MdMessage) => {
+      this.messages.toArray().forEach((m:MdMessage) => {
```
Comment (Gregcop1) [inline src/components/form/messages.ts]: why removing this space

File: src/components/form/messages_spec.ts [hunk 4]
```diff
@@ -1,93 +1,134 @@
...
-          return promiseWait().then(() => {
-            let ctrl: Control = (<any>api.container.property).control;
-            expect(ctrl.value).toBe(null);
-            expect(api.container.valid).toBe(false);
-            expect(api.container.valid).toBe(ctrl.valid);
-            ctrl.updateValue('MorTon', {emitEvent: true});
-            api.fixture.detectChanges();
-            expect(api.container.valid).toBe(true);
-          });
+      it('should add md-valid and md-invalid class on [mdMessages] based on field validation state', () => {
```
Comment (Gregcop1) [inline src/components/form/messages_spec.ts]: should be separated tests

File: src/components/form/messages_spec.ts [hunk 5]
```diff
@@ -1,93 +1,134 @@
...
+      });
+
+      it('should re-export valid from control or form', () => {
+        return setup().then((api:IFormMessagesFixture) => {
+          api.container.control.updateValue(null, {emitEvent: true});
+          api.fixture.detectChanges();
+          expect(api.container.control.value).toBe(null);
+          expect(api.container.valid).toBe(false);
+          expect(api.container.formDirective.valid).toBe(false);
+          api.container.control.updateValue('MorTon', {emitEvent: true});
```
Comment (Gregcop1) [inline src/components/form/messages_spec.ts]: One it per test. This test is build to detect change so it shouldn't test the default state. If you want to test default state, it should be a test only for this

File: modules/site/src/app/examples/input/input-basic-usage.component.ts (hunk 1 above)
Comment (vincentpalita) [inline modules/site/src/app/examples/input/input-basic-usage.component.ts]: I didn't touch that part. My IDE just did some code reformat that's why the line have changed.

As this is unused code, I propose to comment it for now.

File: modules/site/src/main.ts (hunk 2 above)
Comment (vincentpalita) [inline modules/site/src/main.ts]: You are right, switching to local.

File: src/components/form/messages.ts (hunk 3 above)
Comment (vincentpalita) [inline src/components/form/messages.ts]: Sorry, IDE reformat

**Comment by vincentpalita** [comment]
Updated the PR based on your code review @Gregcop1! Thank you for :)

File: modules/site/src/app/examples/input/input-basic-usage.component.ts (hunk 1 above)
Comment (Gregcop1) [inline modules/site/src/app/examples/input/input-basic-usage.component.ts]: Don't comment. Remove it or not but keeping comment code on a GIT project is kind of non-sense

**Comment by Gregcop1** [comment]
Waiting for #279

**Comment by justindujardin** [comment]
@vincentpalita this looks really nice, thanks for the contribution.  

Some of the tests (specifically the ones about md-valid/md-invalid) are failing on some browsers. It looks like the classname is not added in a consistently synchronous way across browsers.  

Here's the failing test output:

```
FAILED TESTS:
  Form Messages
    md-messages
      ✖ should add md-invalid class on [mdMessages] based on field validation state
        IE 11.0.0 (Windows 10 0.0.0)
      Expected false to be truthy.
      Error: Expected false to be truthy.
         at Anonymous function (eval code:103:25)
         at ZoneDelegate.prototype.invoke (/ng2-material/node_modules/zone.js/dist/zone.js:321:14)
         at Zone.prototype.run (/ng2-material/node_modules/zone.js/dist/zone.js:216:18)
         at Anonymous function (/ng2-material/node_modules/zone.js/dist/zone.js:571:18)
         at ZoneDelegate.prototype.invokeTask (/ng2-material/node_modules/zone.js/dist/zone.js:354:18)
         at Zone.prototype.runTask (/ng2-material/node_modules/zone.js/dist/zone.js:256:22)
         at invoke (/ng2-material/node_modules/zone.js/dist/zone.js:423:22)
         at FakeAsyncTestZoneSpec.prototype.flushMicrotasks (/ng2-material/node_modules/zone.js/dist/fake-async-test.js:225:18)
         at flushMicrotasks (eval code:97:5)
         at Anonymous function (eval code:42:13)
         at ZoneDelegate.prototype.invoke (/ng2-material/node_modules/zone.js/dist/zone.js:321:14)
         at Zone.prototype.run (/ng2-material/node_modules/zone.js/dist/zone.js:216:18)
         at Anonymous function (eval code:40:9)

      ✖ should add md-valid class on [mdMessages] based on field validation state
        IE 11.0.0 (Windows 10 0.0.0)
      Expected null not to be null.
      Error: Expected null not to be null.
         at Anonymous function (eval code:120:25)
         at ZoneDelegate.prototype.invoke (/ng2-material/node_modules/zone.js/dist/zone.js:321:14)
         at Zone.prototype.run (/ng2-material/node_modules/zone.js/dist/zone.js:216:18)
         at Anonymous function (/ng2-material/node_modules/zone.js/dist/zone.js:571:18)
         at ZoneDelegate.prototype.invokeTask (/ng2-material/node_modules/zone.js/dist/zone.js:354:18)
         at Zone.prototype.runTask (/ng2-material/node_modules/zone.js/dist/zone.js:256:22)
         at invoke (/ng2-material/node_modules/zone.js/dist/zone.js:423:22)
         at FakeAsyncTestZoneSpec.prototype.flushMicrotasks (/ng2-material/node_modules/zone.js/dist/fake-async-test.js:225:18)
         at flushMicrotasks (eval code:97:5)
         at Anonymous function (eval code:42:13)
         at ZoneDelegate.prototype.invoke (/ng2-material/node_modules/zone.js/dist/zone.js:321:14)
         at Zone.prototype.run (/ng2-material/node_modules/zone.js/dist/zone.js:216:18)
         at Anonymous function (eval code:40:9)
```

I wonder if it has something to do with this PR https://github.com/angular/angular/issues/10148 ?

**Comment by vincentpalita** [comment]
Hi @justindujardin! Thank you for your message. 

I have been on holidays for a few days and didn't have a chance to check that before. 

Will try to take a look during the weekend!

Cheers
Vincent

**Comment by vincentpalita** [comment]
Hi @justindujardin,

I cannot get my environment working on Windows and I am having quite a hard time to invest more time right now on this.
I tested to include the test in the whenStable method but it still doesn't work.

Do you have a running Windows environment where you could help me fix this?
Thank you,
Vincent

**Pull request closed without merging**